perma_web/perma/models.py

```diff
@@ -1230,6 +1230,76 @@ def clear_cache(self):
     def accessible_to(self, user):
         return user.can_edit(self)
 
+    ###
+    ### Methods for playback via Webrecorder
+    ###
+
+    @cached_property
+    def wr_collection_slug(self):
+        return self.guid.lower()
+
+    def wr_iframe_prefix(self, wr_username):
+        return "{}/{}/{}/".format(settings.PLAYBACK_HOST, wr_username, self.wr_collection_slug)
+
+    def init_replay_for_user(self, request):
+        wr_username, wr_session_cookie, is_new_session = get_wr_session(request)
+        if is_new_session or not get_wr_uploaded(request, self.wr_collection_slug):
+            try:
+                self.upload_to_wr(wr_username, wr_session_cookie)
+                set_wr_uploaded(request, self.wr_collection_slug)
+            except WebrecorderException:
+                clear_wr_session(request)
+                raise
+        return wr_username
+
+    def upload_to_wr(self, wr_username, wr_session_cookie):
+        """
+        Upload warc to a temporary Webrecorder per-user collection for playback
+        (The collection is unique per user and per GUID)
+        """
+
+        # Create the temporary collection
+        query_wr_api(
+            method='post',
+            path='/collections?user={user}'.format(user=wr_username),
+            json={'title': self.wr_collection_slug, 'external': True},
+            cookie=wr_session_cookie,
+            valid_if=lambda code, data: code == 200 or (code == 400 and data == {'error': 'duplicate_name'})
+        )
+
+        warc_path = self.warc_storage_file()
+        upload_data = None
+        start_time = time.time()
+
+        with default_storage.open(warc_path, 'rb') as warc_file:
+            _, upload_data = query_wr_api(
+                method='put',
+                path='/upload?force-coll={coll}&filename={coll}.warc.gz'.format(coll=self.wr_collection_slug),
+                data=warc_file,
+                cookie=wr_session_cookie,
+                valid_if=lambda code, data: code == 200 and data.get('upload_id')
+            )
+
+        # wait for WR to finish uploading the WARC
+        while True:
+            try:
+                query_wr_api(
+                    method='get',
+                    path='/upload/{upload_id}?user={user}'.format(user=wr_username, upload_id=upload_data.get('upload_id')),
+                    cookie=wr_session_cookie,
+                    valid_if=lambda code, data: code == 200 and data.get('done'))
+
+                # successfully uploaded if reached here
+                return
+
+            except WebrecorderException as e:
+                time.sleep(0.5)
+                # if all uploaded (size == total_size), but not yet done after
+                # timeout, likely something is wrong, so raise exception
+                if ((upload_data.get('total_size') == upload_data.get('size')) and
```

Comment:
A question about this... `upload_data` is not updated during this `while True` loop, but rather is set once at line 1275 by the return value of the `PUT` call. Does that mean there are two separate error conditions here: one where PUT returns `upload_data.get('total_size') != upload_data.get('size')`, and one where `upload_data.get('total_size') == upload_data.get('size')`, but the `GET` never succeeds?

Reply:
Good catch! The `upload_data` should be getting updated in the loop, now fixed.
The api returns a json with several fields, including `size` / `total_size` and `done` boolean.
The size/total_size is useful if displaying a graphical progress bar.
However, the `done` field indicates that the upload finished and succeeded.

Occasionally, we have seen an error case where the WARC has uploaded, (size == total_size), but `done` is not set to true. This indicates some kind of error (which could be reported in a better as well).
For now, if the size == total_size but done is still not true after some time, treat this as an upload failure.

Reply:
Or, to put another way, if the WARC has been fully uploaded/sent, but WR has not returned that its done after the timeout, then it really has failed to fully upload.